runtime/common/src/oracle.rs

```diff
@@ -0,0 +1,144 @@
+use cfg_primitives::{
+	conversion::fixed_point_to_balance,
+	types::{AccountId, Balance, Moment, PoolId},
+};
+use cfg_traits::PoolInspect;
+use cfg_types::{
+	fixed_point::Quantity,
+	oracles::OracleKey,
+	tokens::{CurrencyId, CustomMetadata},
+};
+use orml_traits::{asset_registry, CombineData, DataProviderExtended, OnNewData};
+use sp_runtime::traits::Zero;
+use sp_std::{marker::PhantomData, vec::Vec};
+
+type TimestampedQuantity = orml_oracle::TimestampedValue<Quantity, Moment>;
+
+/// A provider that maps an `TimestampedQuantity` into a tuple
+/// `(Balance, Moment)`.
+pub struct DataProviderBridge<Oracle, AssetRegistry, Pools>(
+	PhantomData<(Oracle, AssetRegistry, Pools)>,
+);
+
+impl<Oracle, AssetRegistry, Pools> DataProviderExtended<(OracleKey, PoolId), (Balance, Moment)>
+	for DataProviderBridge<Oracle, AssetRegistry, Pools>
+where
+	Oracle: DataProviderExtended<OracleKey, TimestampedQuantity>,
+	AssetRegistry: asset_registry::Inspect<AssetId = CurrencyId, CustomMetadata = CustomMetadata>,
+	Pools: PoolInspect<AccountId, CurrencyId, PoolId = PoolId>,
+{
+	fn get_no_op((key, pool_id): &(OracleKey, PoolId)) -> Option<(Balance, Moment)> {
+		let TimestampedQuantity { value, timestamp } = Oracle::get_no_op(key)?;
+		let currency = Pools::currency_for(*pool_id)?;
+		let decimals = AssetRegistry::metadata(&currency)?.decimals;
+
+		let balance = fixed_point_to_balance(value, decimals as usize).ok()?;
+
+		Some((balance, timestamp))
+	}
+
+	fn get_all_values() -> Vec<((OracleKey, PoolId), Option<(Balance, Moment)>)> {
+		// Unimplemented.
```

Comment:
could we use the `unimplemented!("")` macro here instead or could that cause issue down the line?

Reply:
I wanted to do so, but I can not statically prove that another entity call tomorrow to that method, crashing the node 😞

Reply:
Yeah precisely my concern. This is fine 👍